File: solutions/2026-08-03/python/2026-08-03-stone-game-iii.py
```python
# Summary: Alice and Bob take turns taking 1, 2, or 3 stones from the beginning of a row.
# The player with the highest total score wins. Both play optimally.
# Link: https://leetcode.com/problems/stone-game-iii/
#
# Approach:
# This problem can be solved using dynamic programming. Let dp[i] represent the maximum
# *difference* in score the current player can achieve when considering the stones
# from index i to the end of the array. The current player wants to maximize their score
# minus the opponent's score.
#
# Base cases:
# If i is out of bounds (i >= n), the difference is 0.
#
# Transitions:
# For each index i, the current player can choose to take 1, 2, or 3 stones.
# If the player takes k stones (where k is 1, 2, or 3), their score increases by
# sum(stoneValue[i:i+k]). The opponent will then play optimally on the remaining stones
# starting from index i+k. The opponent's maximum difference from their perspective
# on the remaining stones will be dp[i+k]. Since the opponent is trying to maximize
# their score minus the current player's score on the remaining stones, the current
# player's score difference from this move will be sum(stoneValue[i:i+k]) - dp[i+k].
#
# The current player will choose the move (k=1, 2, or 3) that maximizes this difference.
# dp[i] = max(
#     stoneValue[i] - dp[i+1],  # Take 1 stone
#     stoneValue[i] + stoneValue[i+1] - dp[i+2],  # Take 2 stones
#     stoneValue[i] + stoneValue[i+1] + stoneValue[i+2] - dp[i+3]  # Take 3 stones
# )
# We need to handle boundary conditions where i+k goes beyond the array length.
#
# To implement this efficiently, we can use a suffix sum array to quickly calculate
# the sum of stones taken. Let prefixSum[i] be the sum of stones from index 0 to i-1.
# Then sum(stoneValue[i:j]) = prefixSum[j] - prefixSum[i].
#
# The DP state `dp[i]` will store the maximum score difference the current player
# can achieve starting from `stones[i:]`.
# We iterate from the end of the stones array backwards to compute the DP table.
#
# The final answer is determined by `dp[0]`.
# If `dp[0] > 0`, Alice wins.
# If `dp[0] < 0`, Bob wins.
# If `dp[0] == 0`, it's a tie.
#
# Time Complexity: O(n), where n is the number of stones. We iterate through the stones array once
# to compute the prefix sums and once again to compute the DP table. Each DP transition takes constant time.
# Space Complexity: O(n) for the DP table and the prefix sum array.
class Solution:
    def stoneGameIII(self, stoneValue: list[int]) -> str:
        n = len(stoneValue)

        # Calculate prefix sums for efficient stone sum calculation.
        # prefixSum[i] will store the sum of stoneValue[0] to stoneValue[i-1].
        # prefixSum[0] = 0
        # prefixSum[i] = stoneValue[0] + ... + stoneValue[i-1] for i > 0
        prefixSum = [0] * (n + 1)
        for i in range(n):
            prefixSum[i + 1] = prefixSum[i] + stoneValue[i]

        # dp[i] will store the maximum score difference the current player can achieve
        # when considering the stones from index i to the end (i.e., stoneValue[i:]).
        # The difference is current player's score - opponent's score.
        dp = [0] * (n + 1)

        # Iterate backwards from the end of the stones array.
        # dp[n] is 0 because there are no stones left.
        for i in range(n - 1, -1, -1):
            # For each position i, the current player can take 1, 2, or 3 stones.
            # The player wants to maximize their score difference.
            # The opponent will then play optimally on the remaining stones.

            # Option 1: Take 1 stone
            # Current player's score: stoneValue[i]
            # Opponent's max score difference on remaining stones: dp[i+1]
            # Current player's total difference: stoneValue[i] - dp[i+1]
            take1 = stoneValue[i] - dp[i + 1]

            # Option 2: Take 2 stones (if possible)
            # Current player's score: stoneValue[i] + stoneValue[i+1]
            # Opponent's max score difference on remaining stones: dp[i+2]
            # Current player's total difference: (stoneValue[i] + stoneValue[i+1]) - dp[i+2]
            take2 = float('-inf') # Initialize to negative infinity
            if i + 2 <= n:
                take2 = (stoneValue[i] + stoneValue[i + 1]) - dp[i + 2]

            # Option 3: Take 3 stones (if possible)
            # Current player's score: stoneValue[i] + stoneValue[i+1] + stoneValue[i+2]
            # Opponent's max score difference on remaining stones: dp[i+3]
            # Current player's total difference: (stoneValue[i] + stoneValue[i+1] + stoneValue[i+2]) - dp[i+3]
            take3 = float('-inf') # Initialize to negative infinity
            if i + 3 <= n:
                take3 = (stoneValue[i] + stoneValue[i + 1] + stoneValue[i + 2]) - dp[i + 3]

            # The current player chooses the move that maximizes their score difference.
            dp[i] = max(take1, take2, take3)

        # dp[0] represents the maximum score difference Alice can achieve starting from the beginning.
        # If dp[0] > 0, Alice's score is greater than Bob's.
        # If dp[0] < 0, Bob's score is greater than Alice's.
        # If dp[0] == 0, their scores are equal.
        if dp[0] > 0:
            return "Alice"
        elif dp[0] < 0:
            return "Bob"
        else:
            return "Tie"

```

File: solutions/2026-08-03/javascript/2026-08-03-stone-game-iii.js
/**
 * @fileoverview LeetCode problem "Stone Game III".
 * Problem Summary: Alice and Bob play a game taking stones from a row. Each turn, a player can take 1, 2, or 3 stones from the beginning of the row. The player with the highest score wins. Alice goes first. Both play optimally.
 * Link: https://leetcode.com/problems/stone-game-iii/
 *
 * Approach:
 * This problem can be solved using dynamic programming. We are looking for the maximum score difference a player can achieve given the remaining stones.
 * Let dp[i] represent the maximum score difference the current player can achieve if the game starts with stones from index `i` to the end.
 *
 * The total sum of stones from index `i` to the end is `suffixSum[i]`.
 * If the current player takes `k` stones (where `k` is 1, 2, or 3), they gain `stones[i] + ... + stones[i+k-1]`.
 * The remaining stones start from index `i+k`. The *next* player will then play optimally on `stones[i+k:]`.
 * The score difference for the current player, if they take `k` stones, will be:
 * (sum of stones taken by current player) - (maximum score difference the *next* player can achieve from `stones[i+k:]`)
 *
 * So, for `dp[i]`, the current player will choose `k` (1, 2, or 3) to maximize their score difference:
 * `dp[i] = max(
 *     (stones[i] - dp[i+1]),  // take 1 stone
 *     (stones[i] + stones[i+1] - dp[i+2]), // take 2 stones
 *     (stones[i] + stones[i+1] + stones[i+2] - dp[i+3]) // take 3 stones
 * )`
 *
 * Base cases:
 * If `i` is beyond the array length, `dp[i]` should be 0, as there are no stones left.
 *
 * We iterate backwards from the end of the array to compute `dp` values.
 * `dp[n]` (where `n` is `stoneValue.length`) will be 0.
 * `dp[n-1]`: The player can only take 1 stone. `dp[n-1] = stones[n-1] - dp[n] = stones[n-1]`.
 * `dp[n-2]`: The player can take 1 or 2 stones.
 *   - Take 1: `stones[n-2] - dp[n-1]`
 *   - Take 2: `stones[n-2] + stones[n-1] - dp[n]`
 *   `dp[n-2] = max(stones[n-2] - dp[n-1], stones[n-2] + stones[n-1])`
 *
 * To handle boundary conditions for `dp[i+1]`, `dp[i+2]`, `dp[i+3]`, we can pad the `dp` array or use conditional checks. A simpler approach is to use a `suffixSum` array.
 * Let `suffixSum[i]` be the sum of `stoneValue[i:]`.
 * The sum of stones taken if we take `k` stones from index `i` is `suffixSum[i] - suffixSum[i+k]`.
 *
 * So, the recurrence becomes:
 * `dp[i] = max(
 *     (suffixSum[i] - suffixSum[i+1]) - dp[i+1], // take 1 stone
 *     (suffixSum[i] - suffixSum[i+2]) - dp[i+2], // take 2 stones
 *     (suffixSum[i] - suffixSum[i+3]) - dp[i+3]  // take 3 stones
 * )`
 *
 * Where `suffixSum[n] = 0`, `suffixSum[n+1] = 0`, `suffixSum[n+2] = 0`.
 * And `dp[n] = 0`, `dp[n+1] = 0`, `dp[n+2] = 0`, `dp[n+3] = 0`.
 *
 * We need to calculate `dp[0]`.
 * If `dp[0] > 0`, Alice wins.
 * If `dp[0] < 0`, Bob wins.
 * If `dp[0] == 0`, it's a tie.
 *
 * Time Complexity Analysis:
 * We iterate through the `stoneValue` array once to compute `suffixSum` (O(n)).
 * We then iterate backwards from `n-1` down to 0 to compute the `dp` array (O(n)).
 * Each `dp` computation takes constant time (max of 3 options).
 * Thus, the overall time complexity is O(n), where n is the number of stones.
 *
 * Space Complexity Analysis:
 * We use a `suffixSum` array of size n+1 (O(n)).
 * We use a `dp` array of size n+4 (to handle boundary cases easily) (O(n)).
 * Thus, the overall space complexity is O(n), where n is the number of stones.
 */

/**
 * @param {number[]} stoneValue
 * @return {string}
 */
var stoneGameIII = function(stoneValue) {
    const n = stoneValue.length;

    // suffixSum[i] stores the sum of stones from index i to the end.
    // We use n+1 length for suffixSum so suffixSum[n] is 0, representing sum from an empty subarray.
    const suffixSum = new Array(n + 1).fill(0);
    for (let i = n - 1; i >= 0; i--) {
        suffixSum[i] = suffixSum[i + 1] + stoneValue[i];
    }

    // dp[i] stores the maximum score difference the current player can achieve
    // if the game starts with stones from index i to the end.
    // We use n+4 length for dp to easily handle cases where i+1, i+2, or i+3 go out of bounds.
    // dp[n], dp[n+1], dp[n+2], dp[n+3] are initialized to 0, as there are no stones left.
    const dp = new Array(n + 4).fill(0);

    // Iterate backwards from the end of the array to compute dp values.
    // dp[i] depends on dp[i+1], dp[i+2], dp[i+3].
    for (let i = n - 1; i >= 0; i--) {
        // The current player can take 1, 2, or 3 stones.
        // For each option, we calculate the score difference:
        // (sum of stones taken) - (max score difference the next player can get from remaining stones)

        let maxDiff = -Infinity; // Initialize with a very small number

        // Option 1: Take 1 stone
        // The sum of stones taken is stoneValue[i].
        // The next player plays on stones from index i+1.
        // The score difference is (stoneValue[i]) - dp[i+1].
        // Or using suffix sums: (suffixSum[i] - suffixSum[i+1]) - dp[i+1]
        maxDiff = Math.max(maxDiff, (suffixSum[i] - suffixSum[i + 1]) - dp[i + 1]);

        // Option 2: Take 2 stones (if possible, i.e., i+1 < n)
        if (i + 1 < n) {
            // The sum of stones taken is stoneValue[i] + stoneValue[i+1].
            // The next player plays on stones from index i+2.
            // The score difference is (stoneValue[i] + stoneValue[i+1]) - dp[i+2].
            // Or using suffix sums: (suffixSum[i] - suffixSum[i+2]) - dp[i+2]
            maxDiff = Math.max(maxDiff, (suffixSum[i] - suffixSum[i + 2]) - dp[i + 2]);
        }

        // Option 3: Take 3 stones (if possible, i.e., i+2 < n)
        if (i + 2 < n) {
            // The sum of stones taken is stoneValue[i] + stoneValue[i+1] + stoneValue[i+2].
            // The next player plays on stones from index i+3.
            // The score difference is (stoneValue[i] + stoneValue[i+1] + stoneValue[i+2]) - dp[i+3].
            // Or using suffix sums: (suffixSum[i] - suffixSum[i+3]) - dp[i+3]
            maxDiff = Math.max(maxDiff, (suffixSum[i] - suffixSum[i + 3]) - dp[i + 3]);
        }

        dp[i] = maxDiff;
    }

    // dp[0] represents the maximum score difference Alice can achieve starting the game.
    const aliceScoreDifference = dp[0];

    if (aliceScoreDifference > 0) {
        return "Alice";
    } else if (aliceScoreDifference < 0) {
        return "Bob";
    } else {
        return "Tie";
    }
};
```